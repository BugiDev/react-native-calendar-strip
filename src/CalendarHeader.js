import React, { Component } from "react";
import PropTypes from "prop-types";
import { Text } from "react-native";

import styles from "./Calendar.style.js";

class CalendarHeader extends Component {
  static propTypes = {
    calendarHeaderFormat: PropTypes.string.isRequired,
    calendarHeaderStyle: PropTypes.oneOfType([
      PropTypes.object,
      PropTypes.number
    ]),
    datesForWeek: PropTypes.array.isRequired
  };

  shouldComponentUpdate(nextProps) {
    return JSON.stringify(this.props) !== JSON.stringify(nextProps);
  }

  //Function that formats the calendar header
  //It also formats the month section if the week is in between months
  formatCalendarHeader(datesForWeek, calendarHeaderFormat) {
    if (!datesForWeek || datesForWeek.length === 0) {
      return "";
    }

    let firstDay = datesForWeek[0];
    let lastDay = datesForWeek[datesForWeek.length - 1];
    let monthFormatting = "";
    //Parsing the month part of the user defined formating
    if ((calendarHeaderFormat.match(/Mo/g) || []).length > 0) {
      monthFormatting = "Mo";
    } else {
      if ((calendarHeaderFormat.match(/M/g) || []).length > 0) {
        for (
          let i = (calendarHeaderFormat.match(/M/g) || []).length;
          i > 0;
          i--
        ) {
          monthFormatting += "M";
        }
      }
    }

    if (firstDay.month() === lastDay.month()) {
      return firstDay.format(calendarHeaderFormat);
    } else if (firstDay.year() !== lastDay.year()) {
      return `${firstDay.format(calendarHeaderFormat)} / ${lastDay.format(
        calendarHeaderFormat
      )}`;
    }

    return `${
      monthFormatting.length > 1 ? firstDay.format(monthFormatting) : ""
    } ${monthFormatting.length > 1 ? "/" : ""} ${lastDay.format(
      calendarHeaderFormat
    )}`;
  }

  render() {
    const headerText = this.formatCalendarHeader(
      this.props.datesForWeek,
      this.props.calendarHeaderFormat
    );
    return (
      <Text
        style={[
          styles.calendarHeader,
          { fontSize: this.props.fontSize },
          this.props.calendarHeaderStyle
        ]}
      >
        {headerText}
      </Text>
    );
  }
}

export default CalendarHeader;
