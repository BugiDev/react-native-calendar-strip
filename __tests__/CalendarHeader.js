import React from "react";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import moment from "moment";

import CalendarHeader from "../src/CalendarHeader";

configure({ adapter: new Adapter() });

const today = moment();

describe("CalendarHeader Component", () => {
  it("should render without issues", () => {
    const component = shallow(
      <CalendarHeader
        calendarHeaderFormat="MMMM YYYY"
        weekStartDate={today}
        weekEndDate={today.clone().add(1, "week")}
        fontSize={20}
        allowHeaderTextScaling={true}
      />
    );

    expect(component).toBeTruthy();
  });

  it("should render custom header without issues", () => {
    const component = shallow(
      <CalendarHeader
        calendarHeaderFormat="MMMM YYYY"
        weekStartDate={today}
        weekEndDate={today.clone().add(1, "week")}
        fontSize={20}
        allowHeaderTextScaling={true}
        headerText={"Custom Header"}
      />
    );

    expect(component).toBeTruthy();
  });
});
