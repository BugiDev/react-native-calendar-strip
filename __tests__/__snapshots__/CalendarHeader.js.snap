// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarHeader Component should render 01 / 02 2018 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "alignSelf": "center",
        "fontSize": 16,
        "fontWeight": "bold",
        "textAlign": "center",
      },
      undefined,
    ]
  }
>
  01 / 02 2018
</Text>
`;

exports[`CalendarHeader Component should render 1st / 2nd 18 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "alignSelf": "center",
        "fontSize": 16,
        "fontWeight": "bold",
        "textAlign": "center",
      },
      undefined,
    ]
  }
>
  1st / 2nd 18
</Text>
`;

exports[`CalendarHeader Component should render December 2017 / January 2018 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "alignSelf": "center",
        "fontSize": 16,
        "fontWeight": "bold",
        "textAlign": "center",
      },
      undefined,
    ]
  }
>
  December 2017 / January 2018
</Text>
`;

exports[`CalendarHeader Component should render December 2017 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "alignSelf": "center",
        "fontSize": 16,
        "fontWeight": "bold",
        "textAlign": "center",
      },
      undefined,
    ]
  }
>
  December 2017
</Text>
`;

exports[`CalendarHeader Component should render Jan / Feb 2018 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "alignSelf": "center",
        "fontSize": 16,
        "fontWeight": "bold",
        "textAlign": "center",
      },
      undefined,
    ]
  }
>
  Jan / Feb 2018
</Text>
`;

exports[`CalendarHeader Component should render January / February 2018 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "alignSelf": "center",
        "fontSize": 16,
        "fontWeight": "bold",
        "textAlign": "center",
      },
      undefined,
    ]
  }
>
  January / February 2018
</Text>
`;

exports[`CalendarHeader Component should render in Comic Sans 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "alignSelf": "center",
        "fontSize": 16,
        "fontWeight": "bold",
        "textAlign": "center",
      },
      Object {
        "fontFamily": "ComicSans",
      },
    ]
  }
>
  December 2017 / January 2018
</Text>
`;

exports[`CalendarHeader Component should render in a red color 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "alignSelf": "center",
        "fontSize": 16,
        "fontWeight": "bold",
        "textAlign": "center",
      },
      Object {
        "color": "red",
      },
    ]
  }
>
  December 2017 / January 2018
</Text>
`;

exports[`CalendarHeader Component should render without issues 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "alignSelf": "center",
        "fontSize": 16,
        "fontWeight": "bold",
        "textAlign": "center",
      },
      undefined,
    ]
  }
>
  December 2017
</Text>
`;
